### PR TITLE
Show bandit data

### PR DESCRIPTION
**Pull request by tomrf1**
Adds an "Analytics" button to the methodologies section if epsilon-greedy is selected. This opens a modal with charts showing impressions and scores per variant.
The data is fetched from Dynamodb by the backend.

Screenshot 2024-12-03 at 15 50 31

Screenshot 2024-11-28 at 15 43 36

### CODE REVIEW

File: app/services/DynamoBanditData.scala [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+case class VariantSummary(variantName: String, mean: Double, views: Double)
+
+case class BanditData(variantSummaries: List[VariantSummary], samples: List[EnrichedTestSampleData])
+object BanditData {
+  implicit val decoder = Decoder[BanditData]
+  implicit val encoder = Encoder[BanditData]
+}
+
+class DynamoBanditData(stage: String, client: DynamoDbClient) extends StrictLogging {
+  private val tableName = s"support-bandit-PROD"
```
Comment (charleycampbell) [inline app/services/DynamoBanditData.scala:44]: What happens on our CODE environment? Do we display PROD data?

Comment (tomrf1) [reply app/services/DynamoBanditData.scala:44]: good spot, I was using PROD data for testing but I'll make this stage-aware

File: app/services/DynamoBanditData.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            ":testName" -> AttributeValue.builder.s(s"${channel}_$testName").build
+          ).asJava)
+          .scanIndexForward(true)
+          .build()
+      ).items()
+    }.mapError(DynamoGetError)
+  }
+
+  private def sampleMean(samples: List[VariantSample], population: Double): Double =
+    samples.foldLeft(0D)((acc, sample) =>
```
[the remark this answers is not in the document]
Comment (tomrf1) [reply app/services/DynamoBanditData.scala:63]: It's actually doing the same thing as this `reduce` in SDC:
https://github.com/guardian/support-dotcom-components/blob/main/src/server/bandit/banditData.ts#L110

`foldLeft` is similar - it iterates through the collection, building up an accumulated value (`acc`) as it goes. The starting value is `0`, and the `D` declares it has type `Double`

[the remark this answers is not in the document]
Comment (charleycampbell) [reply app/services/DynamoBanditData.scala:63]: Ahhh I see! Thanks for the explanation!

**Review by charleycampbell — APPROVED** [review]
This is very cool!!

**Pull request merged**